### PR TITLE
documentation: Clear Containers architecture

### PR DESCRIPTION
**Pull request by sameo**
See https://github.com/01org/cc-oci-runtime/blob/topic/documentation/documentation/architecture.md

Signed-off-by: Samuel Ortiz <sameo@linux.intel.com>

### CODE REVIEW

**Comment by chavafg** [comment]
qa-passed

File: documentation/architecture.md [hunk 1]
```diff
@@ -0,0 +1,247 @@
+# Clear Containers Architecture
```
Comment (grahamwhaley) [inline documentation/architecture.md]: General feedback ... this is a **great** document - one we have had the fragments of scattered around here and there, and were quite over-due pulling it all together in one cohesive place - so, thanks for doing that.
Most comments below are just stylistic or typo nitpicks

File: documentation/architecture.md [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+
+## Overview
+
+This is an architectural overview of Clear Containers, based on the 2.1 release.
+
+The [Clear Containers runtime (cc-oci-runtime)](https://github.com/01org/cc-oci-runtime)
+complies with the [OCI](https://github.com/opencontainers) specifications and thus
+work seamlessly with the [Docker Engine](https://www.docker.com/products/docker-engine)
```
Comment (grahamwhaley) [inline documentation/architecture.md]: work -> works?

File: documentation/architecture.md [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+
+On the host, each container process is reaped by a Docker specific (`containerd-shim`) monitoring
+daemon. As Clear Containers processes run inside their own virtual machines, `containerd-shim`
+can not monitor, control or reap them. `cc-oci-runtime` fixes that issue by creating an
+[additional shim process (`cc-shim`)](https://github.com/01org/cc-oci-runtime/tree/master/shim)
+between `containerd-shim` and `cc-proxy`. A `cc-shim` instance will both forward signals and `stdin`
+streams to the container process on the guest and pass the container `stdout` and `stderr` streams
+back to the Docker engine via `containderd-shim`.
+`cc-oci-runtime` creates a `cc-shim` daemon for each Docker container and for each command Docker
+wants to run withing an already running container (`docker exec`).
```
Comment (jodh-intel) [inline documentation/architecture.md]: Typo: within.

File: documentation/architecture.md [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+Hyperstart opens 2 virtio serial interfaces (Control and I/O) on the guest, and QEMU exposes them
+as serial devices on the host. `cc-oci-runtime` uses the control device for sending container
+management commands to hyperstart while the I/O serial device is used to pass I/O streams (`stdout`,
+`stderr`, `stdin`) between the guest and the Docker Engine.
+
+For any given container, both the init process and all potentially executed commands within that
+container, together with their related I/O streams, need to go through 2 virtio serial interfaces
+exported by QEMU. The [Clear Containers proxy (`cc-proxy`)](https://github.com/01org/cc-oci-runtime/tree/master/proxy)
+multiplexes and demultiplexes those commands and streams for all container virtual machines.
+There is one single `cc-proxy` running instance per Clear Containers host.
```
Comment (jodh-intel) [inline documentation/architecture.md]: I'd write this as, "There is only one `cc-proxy` instance per Clear Containers host."

File: documentation/architecture.md [hunk 5]
```diff
@@ -0,0 +1,247 @@
...
+
+![QEMU/KVM](https://github.com/01org/cc-oci-runtime/blob/topic/documentation/documentation/qemu.png)
+
+#### Hyperstart
+
+`Hyperstart` is a daemon running on the guest as an agent for managing containers and processes
+potentially running within those containers.
+
+It is statically built out of a compact C code base, with a strong focus on both simplicity
+an memory footprint.
```
Comment (jodh-intel) [inline documentation/architecture.md]: Typo: "and".

File: documentation/architecture.md [hunk 6]
```diff
@@ -0,0 +1,247 @@
...
+1. Create the container namespaces (Only the network and mount namespaces are currently supported),
+according to the container OCI configuration file.
+2. Spawn the `cc-shim` process and have it wait on a couple of temporary pipes for:
+   * A `cc-proxy` created file descriptor (one end of a socketpair) for the shim to connect to.
+   * The container `hyperstart` sequence numbers for at most 3 I/O streams (`stdout`, `stderr`, `stdin`). `hyperstart` uses
+   those sequence numbers to multiplex all streams for all processes through one serial interface (The virtio I/O serial
+   one).
+3. Run all the [OCI hooks](https://github.com/opencontainers/runtime-spec/blob/master/config.md#hooks) in the container namespaces,
+as described by the OCI container configuration file.
+4. [Set the container networking up](https://github.com/01org/cc-oci-runtime/blob/topic/documentation/documentation/architecture.md#networking).
```
Comment (jodh-intel) [inline documentation/architecture.md]: It would read slightly better if this was "Set up the container networking".

File: documentation/architecture.md [hunk 7]
```diff
@@ -0,0 +1,247 @@
...
+This must happen after all hooks are done as one of them is potentially setting
+the container networking namespace up.
+5. Create the virtual machine running the container process. The VM `systemd` instance will spawn the `hyperstart` daemon.
+6. Wait for `hyperstart` to signal that it is ready.
+7. Send the pod creation command to `hyperstart`. The `hyperstart` pod is the container process sandbox.
+8. Send the allocateIO command to the proxy, for getting the `hyperstart` I/O sequence numbers described in step 2.
+9. Pass the `cc-proxy` socketpair file descriptor, and the I/O sequence numbers to the listening cc-shim process through the dedicated pipes.
+10. The `cc-shim` instance is put to a stopped state to prevent it from doing I/O before the container is started.
```
Comment (jodh-intel) [inline documentation/architecture.md]: "is put into a stopped ..."

File: documentation/architecture.md [hunk 8]
```diff
@@ -0,0 +1,247 @@
...
+
+![Docker create](https://github.com/01org/cc-oci-runtime/blob/topic/documentation/documentation/create.png)
+
+At that point, the container sandbox is created in the virtual machine and `cc-shim` is stopped on the host.
+But the container process itself is not yet running as one need to call `docker start` for actually starting it.
```
Comment (jodh-intel) [inline documentation/architecture.md]: "as one needs to call `docker start` to actually start it".

File: documentation/architecture.md [hunk 9]
```diff
@@ -0,0 +1,247 @@
...
+The [Clear Containers runtime (cc-oci-runtime)](https://github.com/01org/cc-oci-runtime)
+complies with the [OCI](https://github.com/opencontainers) specifications and thus
+work seamlessly with the [Docker Engine](https://www.docker.com/products/docker-engine)
+pluggable runtime architecture. In other words, one can transparently replace the
+[default Docker runtime (runc)](https://github.com/opencontainers/runc) with `cc-oci-runtime`.
+
+![Docker and Clear Containers](https://github.com/01org/cc-oci-runtime/blob/topic/documentation/documentation/docker-cc.png)
+
+`cc-oci-runtime` creates a QEMU/KVM virtual machine for each container the Docker engine
+tries to create.
```
Comment (jodh-intel) [inline documentation/architecture.md]: I'd be tempted drop the "tries to" bit.

File: documentation/architecture.md [hunk 10]
```diff
@@ -0,0 +1,247 @@
...
+[additional shim process (`cc-shim`)](https://github.com/01org/cc-oci-runtime/tree/master/shim)
+between `containerd-shim` and `cc-proxy`. A `cc-shim` instance will both forward signals and `stdin`
+streams to the container process on the guest and pass the container `stdout` and `stderr` streams
+back to the Docker engine via `containderd-shim`.
+`cc-oci-runtime` creates a `cc-shim` daemon for each Docker container and for each command Docker
+wants to run withing an already running container (`docker exec`).
+
+The container workload, i.e. the actual OCI bundle rootfs, is exported from the host to
+the virtual machine via a 9pfs virtio mount point. Hyperstart uses this mount point as the
+container process root filesystem.
```
Comment (jodh-intel) [inline documentation/architecture.md]: "as the root filesystem for the container processes".

File: documentation/architecture.md [hunk 11]
```diff
@@ -0,0 +1,247 @@
...
+	Fsmap         []*FsmapDescriptor  `json:"fsmap,omitempty"`
+	Sysctl        map[string]string   `json:"sysctl,omitempty"`
+	Process       *Process            `json:"process"`
+	RestartPolicy string              `json:"restartPolicy"`
+	Initialize    bool                `json:"initialize"`
+	Ports         []Port              `json:"ports,omitempty"` //deprecated
+}
+```
+
+`Hyperstart` uses a separate serial channel for passing container processes output streams
```
Comment (jodh-intel) [inline documentation/architecture.md]: "for passing the container processes".

File: documentation/architecture.md [hunk 12]
```diff
@@ -0,0 +1,247 @@
...
+
+This is an architectural overview of Clear Containers, based on the 2.1 release.
+
+The [Clear Containers runtime (cc-oci-runtime)](https://github.com/01org/cc-oci-runtime)
+complies with the [OCI](https://github.com/opencontainers) specifications and thus
+work seamlessly with the [Docker Engine](https://www.docker.com/products/docker-engine)
+pluggable runtime architecture. In other words, one can transparently replace the
+[default Docker runtime (runc)](https://github.com/opencontainers/runc) with `cc-oci-runtime`.
+
+![Docker and Clear Containers](https://github.com/01org/cc-oci-runtime/blob/topic/documentation/documentation/docker-cc.png)
```
Comment (grahamwhaley) [inline documentation/architecture.md]: oh, and it would be excellent if we could have the 'source' for the images commit as well, to help us make future edits etc.

File: documentation/architecture.md [hunk 13]
```diff
@@ -0,0 +1,247 @@
...
+10. The `cc-shim` instance is put to a stopped state to prevent it from doing I/O before the container is started.
+
+![Docker create](https://github.com/01org/cc-oci-runtime/blob/topic/documentation/documentation/create.png)
+
+At that point, the container sandbox is created in the virtual machine and `cc-shim` is stopped on the host.
+But the container process itself is not yet running as one need to call `docker start` for actually starting it.
+
+##### `start`
+
+On namespace containers `start` mostly launches the container process on its namespaces.
```
Comment (jodh-intel) [inline documentation/architecture.md]: "`start` launches a traditional Linux container process in its own set of namespaces."

File: documentation/architecture.md [hunk 14]
```diff
@@ -0,0 +1,247 @@
...
+With Clear Containers, `cc-oci-runtime` main task is to create and start a container within the
```
Comment (jodh-intel) [inline documentation/architecture.md]: "the main task of `cc-oci-runtime` is to..."

File: documentation/architecture.md [hunk 15]
```diff
@@ -0,0 +1,247 @@
...
+pod that got created during the `create` step. In practice, this means `cc-oci-runtime` follow
```
Comment (jodh-intel) [inline documentation/architecture.md]: s/follow/follows/

File: documentation/architecture.md [hunk 16]
```diff
@@ -0,0 +1,247 @@
...
+
+1. `cc-oci-runtime` connects to `cc-proxy` and sends it the `attach` command to let it know
+on which pod the container it tries to `kill` is running.
+2. `cc-oci-runtime` sends an hyperstart `KILLCONTAINER` command to `kill` the container running
+on the guest. The command is sent to `cc-proxy` who forwards it to the right hyperstart instance
+running in the appropriate guest.
+
+##### `delete`
+
+`docker delete` is about deleting all resources help by a stopped/killed container. Running
```
Comment (jodh-intel) [inline documentation/architecture.md]: s/help/held/

File: documentation/architecture.md [hunk 17]
```diff
@@ -0,0 +1,247 @@
...
+
+#### Proxy
+
+#### Shim
+
+#### Networking
+
+Containers will typically live in their own, possibly shared, networking namespace.
+At some point in a container lifecycle, container engines will set that namespace
+up to add the container to an isolated but shared across containers network.
```
Comment (jodh-intel) [inline documentation/architecture.md]: "will set up that namespace to add the container to a network which is isolated from the host network, but which is shared between containers" ?

File: documentation/architecture.md [hunk 18]
```diff
@@ -0,0 +1,247 @@
...
+
+In order to do so, container engines will usually add one end of a `virtual ethernet
+(veth)` pair into the container networking namespace. The other end of the `veth` pair
+is added to the container network.
+
+This is a very namespace centric approach as QEMU can not handle `veth` interfaces.
```
Comment (jodh-intel) [inline documentation/architecture.md]: s/namespace centric/namespace-centric/

File: documentation/architecture.md [hunk 19]
```diff
@@ -0,0 +1,247 @@
...
+2. `cc-oci-runtime` sends an hyperstart `DESTROYPOD` command to `destroy` the pod holding the
+container running on the guest. The command is sent to `cc-proxy` who forwards it to the right
+hyperstart instance running in the appropriate guest.
+3. After deleting the last running pod, `hyperstart` will gracefully shut the virtual machine
+down.
+4. `cc-oci-runtime` sends the `BYE` command to `cc-proxy`, to let it know that a given virtual
+machine shut down. `cc-proxy` will then clean all its internal resources associated with this
+VM.
+
+#### Proxy
```
Comment (jodh-intel) [inline documentation/architecture.md]: I guess this section and the section below are to complete later?

File: documentation/architecture.md (hunk 1 above)
Comment (jodh-intel) [inline documentation/architecture.md]: +1 - this is superb. Just some minor nits.

We do need to either update or just remove the following which is (now) 2.0-specific (missing all the new 2.1 goodness ;-):

- https://github.com/01org/cc-oci-runtime/blob/master/DESIGN.rst#code-flow-with-docker-1-12

File: documentation/architecture.md (hunk 19 above)
Comment (sameo) [inline documentation/architecture.md]: They are complete now.

**Comment by sameo** [comment]
@jodh-intel @grahamwhaley Thanks for the review, much appreciated.

**Comment by sameo** [comment]
This document is complete now, ready to be reviewed/merged.

File: documentation/architecture.md [hunk 20]
```diff
@@ -0,0 +1,336 @@
...
+we want to use to run the `exec` command.
+2. `cc-oci-runtime` sends the allocateIO command to the proxy, for getting the `hyperstart` I/O sequence
+numbers for the `exec` command I/O streams.
+3. `cc-oci-runtime` sends an hyperstart `EXECMD` command to start the command in the right container
+The command is sent to `cc-proxy` who forwards it to the right hyperstart instance running
+in the appropriate guest.
+4. Spawn the `cc-shim` process for it to forward the output streams (stderr and stdout) and the `exec`
+command exit code to Docker.
+
+Now the Clear Container is running as an isolated process in its own virtual machine.
```
Comment (grahamwhaley) [inline documentation/architecture.md]: As we are talking about `exec` here, I think this sentence should be talking about how the exec process is running in, maybe?, its own namespace - or, is now running in the same namespace etc. (that is, describe what is different/the same between the original workload and an exec'd one within the same container).

I suspect this sentence may either be a copy/paste hangover, or it should be more detached from the exec section ?

File: documentation/architecture.md [hunk 21]
```diff
@@ -0,0 +1,336 @@
...
+
+##### `kill`
+
+When sending the OCI `kill` command, container runtimes should send a [UNIX signal](https://en.wikipedia.org/wiki/Unix_signal)
+to the container process.
+In the Clear Containers context, this means `cc-oci-runtime` needs a way to send a signal
+to the container process within the virtual machine. As `cc-shim` is responsible for
+forwarding signal to its associated running containers, `cc-oci-runtime` naturally
```
Comment (grahamwhaley) [inline documentation/architecture.md]: s/signal/signals/

File: documentation/architecture.md [hunk 22]
```diff
@@ -0,0 +1,336 @@
...
+calls `kill` on the `cc-shim` PID.
+
+However, `cc-shim` is not able to trap `SIGKILL` and `SIGSTOP` and thus `cc-oci-runtime`
+needs to follow a different code path for those 2 signals.
+Instead of `kill`'ing the `cc-shim` PID, it will go through the following steps:
+
+1. `cc-oci-runtime` connects to `cc-proxy` and sends it the `attach` command to let it know
+on which pod the container it tries to `kill` is running.
```
Comment (grahamwhaley) [inline documentation/architecture.md]: s/it tries/it is trying/

File: documentation/architecture.md [hunk 23]
```diff
@@ -0,0 +1,336 @@
...
+containers can not be `delete`d unless the OCI runtime is explictly being asked to. In that
+case it will first `kill` the container and only then `delete` it.
+
+The resources held by a Clear Container are quite different from the ones held by a host
+namespace container e.g. run by `runc`. `cc-oci-runtime` needs mostly to delete the pod
+holding the stopped container on the virtual machine, shut the hypervisor down and finally
+delete all related proxy resources:
+
+1. `cc-oci-runtime` connects to `cc-proxy` and sends it the `attach` command to let it know
+on which pod the container it tries to `delete` is running.
```
Comment (grahamwhaley) [inline documentation/architecture.md]: s/it tries/it is trying/

File: documentation/architecture.md [hunk 24]
```diff
@@ -0,0 +1,336 @@
...
+2. `cc-oci-runtime` sends an hyperstart `DESTROYPOD` command to `destroy` the pod holding the
+container running on the guest. The command is sent to `cc-proxy` who forwards it to the right
+hyperstart instance running in the appropriate guest.
+3. After deleting the last running pod, `hyperstart` will gracefully shut the virtual machine
+down.
+4. `cc-oci-runtime` sends the `BYE` command to `cc-proxy`, to let it know that a given virtual
+machine shut down. `cc-proxy` will then clean all its internal resources associated with this
```
Comment (grahamwhaley) [inline documentation/architecture.md]: s/shut down/is shut down/  ?

File: documentation/architecture.md [hunk 25]
```diff
@@ -0,0 +1,336 @@
...
+
+- A UNIX, named socket for all `cc-oci-runtime` instances on the host to send commands to `cc-proxy`.
+- One socket pair per `cc-shim` instance, to send stdin and receive stdout and stderr I/O streams. See the
+[cc-shim section](https://github.com/01org/cc-oci-runtime/blob/topic/documentation/documentation/architecture.md#shim)
+for more details about that interface.
+
+The protocol on the `cc-proxy` UNIX named socket supports the following commands:
+
+- `Hello`: This command is for `cc-oci-runtime` to let `cc-proxy` know about a newly created VM that will
+hold containers. This command payload contains the `hyperstart` QEMU exported control and I/O UNIX socket
```
Comment (grahamwhaley) [inline documentation/architecture.md]: I don't understand the phrase 'hyperstart QEMU exported control....' - any way we can clarify this?

File: documentation/architecture.md [hunk 26]
```diff
@@ -0,0 +1,336 @@
...
+paths, and `cc-proxy` will connect to both of them after receiving the `Hello` command.
+- `Bye`: This is the opposite of `Hello`, i.e. `cc-oci-runtime` uses this command to let `cc-proxy` know
+that it can release all resources related to the VM described in the command payload.
+- `Attach`: `cc-oci-runtime` uses that command as VM multiplexer as it allows it to notify `cc-proxy` about
```
Comment (grahamwhaley) [inline documentation/architecture.md]: s/as VM multiplexer/to enable multiplexing multiple VMs'/

File: documentation/architecture.md [hunk 27]
```diff
@@ -0,0 +1,336 @@
...
+which VM it wants to talk to. In other words, this commands allows `cc-oci-runtime` to attach itself to a
+running VM.
+- `AllocateIO`: As `hyperstart` can potentially handle several container processes streams at the same time,
```
Comment (grahamwhaley) [inline documentation/architecture.md]: 'processes streams' I find hard to parse (but I know what it is trying to say). maybe flip it, so more like: 'can handle streams from multiple container processes' ?

File: documentation/architecture.md [hunk 28]
```diff
@@ -0,0 +1,336 @@
...
+it needs to be able to associate any given stream to a container process. This is done by `hyperstart`
+allocating a set of at most 2 so-called sequence number per container process. `cc-oci-runtime` will send
```
Comment (grahamwhaley) [inline documentation/architecture.md]: s/number/numbers/

File: documentation/architecture.md [hunk 29]
```diff
@@ -0,0 +1,336 @@
...
+commands.
+
+For more details about `cc-proxy`'s protocol, theory of operations or debugging tips, please read
+[`cc-proxy` README](https://github.com/01org/cc-oci-runtime/tree/master/proxy).
+
+#### Shim
+
+Docker's `containerd-shim` is designed around the assumption that it can monitor and reap the actual
+container process. As `containerd-shim` runs on the host, it can not directly monitor a process running
+within a virtual machine. At most it can see the QEMU process but that is not enough.
```
Comment (grahamwhaley) [inline documentation/architecture.md]: s/process/process,/

File: documentation/architecture.md [hunk 30]
```diff
@@ -0,0 +1,336 @@
...
+an encapsulation layer between `containerd-shim` and `hyperstart`:
+
+- It fragments and encapsulates the standard input stream from containerd-shim into `hyperstart` stream packets:
+```
+  ┌─────────────────┬───────────────┬────────────────────┐
+  │    IO stream    │ Packet length │ IO stream fragment │
+  │ sequence number │    (32 bits)  │                    │
+  └─────────────────┴───────────────┴────────────────────┘
+```
+- It deencapsulates and assemble standard output and error `hyperstart` stream packets into an output stream
```
Comment (grahamwhaley) [inline documentation/architecture.md]: s/deencapsulates/de-encapsulates/

Comment (grahamwhaley) [inline documentation/architecture.md]: s/assemble/assembles/

File: documentation/architecture.md [hunk 31]
```diff
@@ -0,0 +1,336 @@
...
+that it forwards to `containerd-shim`
+- It translates all UNIX signals (except `SIGKILL` and `SIGSTOP`) into `hyperstart` ``KILLCONTAINER` commands
```
Comment (grahamwhaley) [inline documentation/architecture.md]: missing close quote after KILLCONTAINER (obvious in cooked mode during review)

File: documentation/architecture.md [hunk 32]
```diff
@@ -0,0 +1,336 @@
...
+In order to do so, container engines will usually add one end of a `virtual ethernet
+(veth)` pair into the container networking namespace. The other end of the `veth` pair
+is added to the container network.
+
+This is a very namespace-centric approach as QEMU can not handle `veth` interfaces.
+Instead it typically creates `TAP` interfaces for adding connectivity to a virtual
+machine.
+
+To overcome that incompatibility between typical container engines expectations
+and virtual machines, `cc-oci-runtime` networking implementation intents to transparently
```
Comment (grahamwhaley) [inline documentation/architecture.md]: 'intents'. Not sure that is the right word here - but struggling to find a suitable one word replacement. Hmm, maybe 'implements' ?

File: documentation/architecture.md [hunk 33]
```diff
@@ -0,0 +1,337 @@
...
+3. After deleting the last running pod, `hyperstart` will gracefully shut the virtual machine
+down.
+4. `cc-oci-runtime` sends the `BYE` command to `cc-proxy`, to let it know that a given virtual
+machine is shut down. `cc-proxy` will then clean all its internal resources associated with this
+VM.
+
+#### Proxy
+
+`cc-proxy` is a daemon offering access to the [`hyperstart`](https://github.com/hyperhq/hyperstart)
+VM agent to multiple clients like for example `cc-shim` and `cc-proxy`.
```
Comment (jodh-intel) [inline documentation/architecture.md]: This sentence defines `cc-proxy` by referring to `cc-proxy`. I think the following change captures the minimal definition:

"VM agent to multiple `cc-shim` clients."

File: documentation/architecture.md [hunk 34]
```diff
@@ -0,0 +1,337 @@
...
+- It de-encapsulates and assembles standard output and error `hyperstart` stream packets into an output stream
+that it forwards to `containerd-shim`
+- It translates all UNIX signals (except `SIGKILL` and `SIGSTOP`) into `hyperstart` `KILLCONTAINER` commands
+that it sends to the VM via `cc-proxy` UNIX named socket.
+
+The IO stream sequence numbers are passed from `cc-proxy` to `cc-shim` when the former spawns the latter.
+They are generated by `hyperstart` and `cc-oci-runtime` fetches them by sending the `AllocateIO` command to
+`cc-proxy`.
+
+As an example, let's say that running the `# pwd` command from a container standard input will generate
```
Comment (jodh-intel) [inline documentation/architecture.md]: I'd drop the prompt and make this simply: "the `pwd` command".

File: documentation/architecture.md [hunk 35]
```diff
@@ -0,0 +1,337 @@
...
+`/tmp` from the container standard output. `hyperstart` assigned this specific process 8888 and 8889 respectively
+as the stdin, stdour and stderr sequence numbers.
+With `cc-shim` and Clear Containers, this example would look like:
+
+![cc-shim](https://github.com/01org/cc-oci-runtime/blob/topic/documentation/documentation/shim.png)
```
Comment (jodh-intel) [inline documentation/architecture.md]: I appreciate this is mentioned in the text, but -- although it might be difficult to squeeze this in -- I think it would help if you could add a visual hint that those 3 boxes are hyperstart packets. Maybe a blue border along with an entry in the key below stdout and stdin would make that clearer (or even just adding "seq", "len" and "data" above each part of the packets).

File: documentation/architecture.md [hunk 36]
```diff
@@ -0,0 +1,337 @@
...
+
+This is an architectural overview of Clear Containers, based on the 2.1 release.
+
+The [Clear Containers runtime (cc-oci-runtime)](https://github.com/01org/cc-oci-runtime)
+complies with the [OCI](https://github.com/opencontainers) specifications and thus
+works seamlessly with the [Docker Engine](https://www.docker.com/products/docker-engine)
+pluggable runtime architecture. In other words, one can transparently replace the
+[default Docker runtime (runc)](https://github.com/opencontainers/runc) with `cc-oci-runtime`.
+
+![Docker and Clear Containers](https://github.com/01org/cc-oci-runtime/blob/topic/documentation/documentation/docker-cc.png)
```
Comment (jodh-intel) [inline documentation/architecture.md]: Would it be possible to add the sources for these graphics files? That would make maintenance easier since without them, we'd probably have to regenerate the diagrams from scratch.

File: documentation/architecture.md [hunk 37]
```diff
@@ -0,0 +1,337 @@
...
+  │    IO stream    │ Packet length │ IO stream fragment │
+  │ sequence number │    (32 bits)  │                    │
+  └─────────────────┴───────────────┴────────────────────┘
+```
+- It de-encapsulates and assembles standard output and error `hyperstart` stream packets into an output stream
+that it forwards to `containerd-shim`
+- It translates all UNIX signals (except `SIGKILL` and `SIGSTOP`) into `hyperstart` `KILLCONTAINER` commands
+that it sends to the VM via `cc-proxy` UNIX named socket.
+
+The IO stream sequence numbers are passed from `cc-proxy` to `cc-shim` when the former spawns the latter.
```
Comment (jodh-intel) [inline documentation/architecture.md]: `cc-oci-runtime` spawns `cc-shim`, not `cc-proxy`.

File: documentation/architecture.md (hunk 35 above)
Comment (jodh-intel) [inline documentation/architecture.md]: Should `cc-shim` be green in this diagram? It does handle `stdin` so maybe that's the reason, but it also handles `stdout`, so could just as well be red. I'm guessing it's green simply to make it different from the other elements (`cc-proxy`, `hyperstart`, `containerd-shim`). If that's the case, I suggest making it a different colour to avoid confusion with "green means `stdin`".

Comment (jodh-intel) [inline documentation/architecture.md]: One more thing - `stdout` is represented by red in the key, but the boxes showing `stdout` data appear to be a different shade of red which is confusing.

**Comment by sameo** [comment]
@jodh-intel Thanks for the second round. All comments should be addressed now.

**Comment by dlespiau** [comment]
lgtm

**Comment by jodh-intel** [comment]
lgtm.

File: documentation/architecture.md [hunk 38]
```diff
@@ -0,0 +1,337 @@
...
+- It translates all UNIX signals (except `SIGKILL` and `SIGSTOP`) into `hyperstart` `KILLCONTAINER` commands
+that it sends to the VM via `cc-proxy` UNIX named socket.
+
+The IO stream sequence numbers are passed from `cc-runtime` to `cc-shim` when the former spawns the latter.
+They are generated by `hyperstart` and `cc-oci-runtime` fetches them by sending the `AllocateIO` command to
+`cc-proxy`.
+
+As an example, let's say that running the `pwd` command from a container standard input will generate
+`/tmp` from the container standard output. `hyperstart` assigned this specific process 8888 and 8889 respectively
+as the stdin, stdour and stderr sequence numbers.
```
Comment (grahamwhaley) [inline documentation/architecture.md]: s/stdour/stdout/

**Comment by grahamwhaley** [comment]
lgtm
